xds/src/test/java/io/grpc/xds/GcpAuthenticationFilterTest.java

```diff
@@ -0,0 +1,159 @@
+
+import com.google.auth.oauth2.ComputeEngineCredentials;
+import com.google.protobuf.Any;
+import com.google.protobuf.Message;
+import com.google.protobuf.UInt64Value;
+import io.envoyproxy.envoy.extensions.filters.http.gcp_authn.v3.GcpAuthnFilterConfig;
+import io.envoyproxy.envoy.extensions.filters.http.gcp_authn.v3.TokenCacheConfig;
+import io.grpc.CallCredentials;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientInterceptor;
+import io.grpc.MethodDescriptor;
+import io.grpc.StringMarshaller;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+@RunWith(JUnit4.class)
+public class GcpAuthenticationFilterTest {
+
+  @Test
+  public void testParseFilterConfig_withValidConfig() {
+    GcpAuthnFilterConfig config = GcpAuthnFilterConfig.newBuilder()
+        .setCacheConfig(TokenCacheConfig.newBuilder().setCacheSize(UInt64Value.of(10)))
+        .build();
+    Any anyMessage = Any.pack(config);
+
+    GcpAuthenticationFilter filter = new GcpAuthenticationFilter();
+    ConfigOrError<? extends Filter.FilterConfig> result = filter.parseFilterConfig(anyMessage);
+
+    assertNotNull(result.config);
+    assertNull(result.errorDetail);
+    assertEquals(10L,
+        ((GcpAuthenticationFilter.GcpAuthenticationConfig)
+            result.config).getCacheSize().longValue());
+  }
+
+  @Test
+  public void testParseFilterConfig_withZeroCacheSize() {
+    GcpAuthnFilterConfig config = GcpAuthnFilterConfig.newBuilder()
+        .setCacheConfig(TokenCacheConfig.newBuilder().setCacheSize(UInt64Value.of(0)))
+        .build();
+    Any anyMessage = Any.pack(config);
+
+    GcpAuthenticationFilter filter = new GcpAuthenticationFilter();
+    ConfigOrError<? extends Filter.FilterConfig> result = filter.parseFilterConfig(anyMessage);
+
+    assertNull(result.config);
+    assertNotNull(result.errorDetail);
+    assertTrue(result.errorDetail.contains("cache_config.cache_size must be in the range"));
+  }
+
+  @Test
+  public void testParseFilterConfig_withInvalidMessageType() {
+    GcpAuthenticationFilter filter = new GcpAuthenticationFilter();
+    Message invalidMessage = Mockito.mock(Message.class);
+    ConfigOrError<? extends Filter.FilterConfig> result = filter.parseFilterConfig(invalidMessage);
+
+    assertNull(result.config);
+    assertNotNull(result.errorDetail);
+    assertTrue(result.errorDetail.contains("Invalid config type"));
+  }
+
+  @Test
+  public void testGetCallCredentials_addsToCache() throws Exception {
+    ComputeEngineCredentials credentials = ComputeEngineCredentials.create();
+    GcpAuthenticationFilter.LruCache<String, CallCredentials> cache =
+        new GcpAuthenticationFilter.LruCache<>(10);
+
+    GcpAuthenticationFilter filter = new GcpAuthenticationFilter();
+    CallCredentials callCredentials =
+        filter.getCallCredentials(cache, "test-audience", credentials);
```

Comment:
I strongly encourage you to use the class's public API when it is possible. This test doesn't actually check that the _filter_ uses the cache. And it could have done so pretty easily by just calling an interceptor twice and comparing the results. Not only does that give us more confidence in the results, but it also is less likely to need changes over time.

Similarly, testLruCacheEviction isn't testing that the filter uses the configured cache size. That test is testing many combinations, so maybe it is still useful to have it look deep into the filter, but preferring testing the API would still have provided value there.

(Those are the two tests I looked at, because I was looking at why the internals were being exposed.)

Reply:
+1 to this. 
"it is still useful to have it look deep into the filter" Actually there is no functionality of the LRU cache that cannot be tested via the filter public APIs. Eviction and new creation of call credentials for an audience can be tested by checking whether filter.getCallCredentials returned the same reference as before or a new one.